File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BooksControllerDeleteBookDataTask.java
package org.nypl.simplified.books.core;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.io7m.jnull.NullCheck;

final class BooksControllerDeleteBookDataTask implements Runnable
{
  private static final Logger        LOG;

  static {
    LOG =
      NullCheck.notNull(LoggerFactory
        .getLogger(BooksControllerDeleteBookDataTask.class));
  }

  private final BookDatabaseType     book_database;
  private final BookID               book_id;
  private final BooksStatusCacheType books_status;

  public BooksControllerDeleteBookDataTask(
    final BooksStatusCacheType in_books_status,
    final BookDatabaseType in_book_database,
    final BookID in_book_id)
  {
    this.books_status = NullCheck.notNull(in_books_status);
    this.book_database = NullCheck.notNull(in_book_database);
    this.book_id = NullCheck.notNull(in_book_id);
  }

  @Override public void run()
  {
    try {
      final BookDatabaseEntryType e =
        this.book_database.getBookDatabaseEntry(this.book_id);
      e.destroyBookData();

      this.books_status.booksStatusUpdate(new BookStatusLoaned(this.book_id));
    } catch (final Throwable e) {
      BooksControllerDeleteBookDataTask.LOG.error(
        "could not destroy book data for {}: ",
        this.book_id,
        e);
    }
  }
}

File: simplified-books-core/src/main/java/org/nypl/simplified/books/core/BooksControllerDeleteBookDataTask.kt
package org.nypl.simplified.books.core

import org.slf4j.LoggerFactory
import java.util.concurrent.Callable

internal class BooksControllerDeleteBookDataTask(
  private val bookStatus: BooksStatusCacheType,
  private val bookDatabase: BookDatabaseType,
  private val bookID: BookID,
  private val needsAuthentication: Boolean) : Callable<Unit> {

  private val log = LoggerFactory.getLogger(BooksControllerDeleteBookDataTask::class.java)

  override fun call() {
    try {
      this.log.debug("[{}]: deleting book data", this.bookID.shortID)

      val databaseEntry = this.bookDatabase.databaseOpenExistingEntry(this.bookID)
      val snap = databaseEntry.entryDeleteBookData()
      val status = BookStatus.fromSnapshot(this.bookID, snap)
      this.bookStatus.booksStatusUpdate(status)

      // destroy entry, this is needed after deletion has been broadcasted, so the book doesn't stay in loans,
      // especially needed for collection without auth requirement where no syn is happening
      if (!this.needsAuthentication) {
        databaseEntry.entryDestroy()
      }
    } catch (e: Throwable) {
      this.log.error("[{}]: could not destroy book data: ", this.bookID.shortID, e)
      throw e
    } finally {
      this.log.debug("[{}]: deletion completed", this.bookID.shortID)
    }
  }
}
